### PR TITLE
Cache the main page resources.

### PR DESCRIPTION
**Pull request by MarkBennett**
This doesn't cache the service worker itself. I also need to explore how cache invalidation works.
- [X] Install a ServiceWorker (worker.js)
- [X] Cache page contents for offline
- [X] Versioned cache with expiry
- [x] Increment cache version as part of `npm run deploy`
- [x] disable cache by default in dev (with a way to enable)
- [x] automatically add new files to cache

### CODE REVIEW

**Comment by MarkBennett** [comment]
It looks like the current best practice is to set a cache version number in the cache name, or in the cache keys and then change this constant in the service worker to refresh page content when desired. It seems a little nasty, but makes sense.

I'll modify this PR to do so shortly. Also adding a manual cache invalidation might make sense.

**Comment by MarkBennett** [comment]
I've got a start on the cache invalidation right now. Just working on some other ways to make it part of the `npm run deploy` script.

File: src/worker.js [hunk 1]
```diff
@@ -1,9 +1,41 @@
+"use scrict"
+
 // The SW will be shutdown when not in use to save memory,
 // be aware that any global state is likely to disappear
 console.log("SW startup");
 
+let CACHE_VERSION = 1;
```
Comment (MarkBennett) [inline src/worker.js]: This is my first pass at versioning. This number is used to create new caches for each increment of this value. Other values are deleted to make room for it.

File: src/worker.js [hunk 2]
```diff
@@ -1,9 +1,41 @@
...
+  return storage.keys().
+    then((keys) => keys.filter(key => key !== cache_name)).
+    then((keys) => keys.map(key => storage.delete(key)));
+};
+
 self.addEventListener('install', function(event) {
   console.log("SW installed");
+
+  openCache(caches, cacheName(CACHE_VERSION)).
+  then(([storage, cache, cache_name]) => cacheRequests(cache).then((_) => [storage, cache_name])).
```
Comment (MarkBennett) [inline src/worker.js]: I'm using a closure to wrap some variables not used by cacheRequest() but used on the following line clearOldCaches(). Anyone aware of a clearer way to express this?

/cc @abramhindle

**Comment by MarkBennett** [comment]
I need to add some tests for the worker, then I think this is done and ready to go.

**Comment by MarkBennett** [comment]
Hey @MHerszak. This is the PR we're using to add ServiceWorkers to the website. Feedback welcome! :)

**Comment by MHerszak** [comment]
@MarkBennett I am not following...

File: index.html [hunk 3]
```diff
@@ -192,7 +192,7 @@
         <div class="col-md-4 col-md-offset-0 col-lg-3 col-lg-offset-1">
           <h3>Where?</h3>
           <p>
-            <img src="https://www.startupedmonton.com/images/logo.png"  class="colOpenGraphic" />
+            <img src="startup-edmonton.png"  class="colOpenGraphic" />
```
Comment (MarkBennett) [inline index.html:195]: This change was included in this branch as I had to list the files the cache in the service worker and realized this was missing.

**Comment by MarkBennett** [comment]
I'm just about ready to merge this but am curious if you guys think I should disable SW caching in dev? This would produce different behaviour from production but would make things less annoying for anyone not actively working on the service worker.

Thoughts @ExchangeJS/organizers?

**Comment by MarkBennett** [comment]
I'm going to look this over again tomorrow, but with the dev flag and actively collecting new files from the file system I think this PR is ready to merge.

**Pull request merged**